src/ngraph/op/one_hot.cpp

```diff
@@ -20,28 +20,58 @@
 using namespace std;
 using namespace ngraph;
 
-op::OneHot::OneHot(const shared_ptr<Node>& arg, const Shape& shape, size_t one_hot_axis)
+op::OneHot::OneHot(const shared_ptr<Node>& arg, const PartialShape& shape, size_t one_hot_axis)
     : Op("OneHot", check_single_output_args({arg}))
     , m_shape(shape)
     , m_one_hot_axis(one_hot_axis)
 {
     constructor_validate_and_infer_types();
+}
+
+void op::OneHot::validate_and_infer_types()
+{
+    element::Type arg_et = get_input_element_type(0);
+    PartialShape arg_shape = get_input_partial_shape(0);
+    Rank arg_rank = arg_shape.rank();
+
+    NODE_VALIDATION_ASSERT(this, m_shape.rank().is_static())
+        << "Requested result shape has dynamic rank.";
+
+    NODE_VALIDATION_ASSERT(this, m_one_hot_axis < size_t(m_shape.rank()))
+        << "One-hot axis (" << m_one_hot_axis
+        << ") is out of bounds (requested result shape: " << m_shape << ").";
+
+    NODE_VALIDATION_ASSERT(this, m_shape[m_one_hot_axis].is_static())
+        << "Requested result shape (" << m_shape << ") has dynamic dimension at the one-hot axis "
+        << "(" << m_one_hot_axis << ").";
 
-    auto& input = m_inputs.at(0);
-    auto& input_element_type = input.get_element_type();
+    PartialShape result_shape{m_shape};
 
-    NODE_VALIDATION_ASSERT(this, one_hot_axis < shape.size())
-        << "One-hot axis (" << one_hot_axis
-        << ") is out of bounds (requested result shape: " << shape << ").";
+    if (arg_rank.is_static())
+    {
+        std::vector<Dimension> expected_input_dims(size_t(m_shape.rank()));
```

Comment:
Indeed, there are so many instances of converting `Dimension`/`Rank` to `size_t` (not just in this PR) that I'm a little concerned using `static_cast` everywhere is going to hurt readability. But I understand that a C-style cast is not desirable. What do you think about adding a function to `Dimension`/`Rank`, so we could have something like the following:

```
std::vector<Dimension> expected_input_size(m_shape.rank().to_size_t());
```

Reply:
FWIW, Google's style guide suggests using a curly-brace initializer when converting between arithmetic types, which would be nice and concise as well:

https://google.github.io/styleguide/cppguide.html#Casting

Reply:
Stroustrup says don't use the `type(x)` for builtin types since it does the `(type)x` thing and there are no safeguards. To use `{}` here you'd need to use a temp rank var.

Reply:
Curly init sounds good.